### DISCUSSION TITLE
Unable to set StyledProperty<bool> to "True"

### DISCUSSION THREAD

**Discussion by KuroiRoy** [AvaloniaUI/Avalonia#10044] [Q&A] [answered]
I've defined a attached property:
``` csharp
public class ScrollViewerExtensionBehaviour : AvaloniaObject {
    public static readonly StyledProperty<bool> ShiftWheelScrollsHorizontallyProperty = 
        AvaloniaProperty.Register<ScrollViewer, bool>(nameof(ShiftWheelScrollsHorizontally));

    public bool ShiftWheelScrollsHorizontally {
        get => GetValue(ShiftWheelScrollsHorizontallyProperty);
        set => SetValue(ShiftWheelScrollsHorizontallyProperty, value);
    }
}
```
And I use it here:
``` xaml
<ScrollViewer extensions:ScrollViewerExtensionBehaviour.ShiftWheelScrollsHorizontally="True">
    <DataGrid Items="{Binding TransactionRecords}" AutoGenerateColumns="True"/>
</ScrollViewer>
```
To me, both the usage and the definition of the property look exactly like the AutoGenerateColumns property on DataGrid. But when I give it the parameter "True" it produces the following error:
```
Import.axaml(15, 19): [XAMLIL] Unable to find suitable setter or adder for property ShiftWheelScrollsHorizontally of type FinanceTracker:FinanceTracker.Extensions.AvaloniaUI.ScrollViewerExtensionBehaviour for argument System.Runtime:System.String, available setter parameter lists are:
Avalonia.UnsetValueType
Avalonia.Data.IBinding Line 15, position 19.
```
Am I missing something?

PS: I just noticed that DataGrid both already has scroll functionality and it can scroll horizontally with shift.
PPS: The documentation only talks about adding properties to your own class or attaching to an existing property on another class. Not about adding a property to a built-in control, so I'm very curious why my implementation doesn't work while the DataGrid one does

**Comment by danielmayost** [comment, marked as answer]
why `StyledProperty`, don't you need `AttachedProperty`?

**Reply by KuroiRoy** [reply]
Simply because that is what the Register function returns and that is what the DataGrid implementation uses

**Reply by danielmayost** [reply]
Maybe I misunderstood, where is this in the DataGrid implementation?

**Reply by KuroiRoy** [reply]
Here, https://github.com/AvaloniaUI/Avalonia/blob/d2bfd61dfcf04ea1436a87452c83ba5ea04e2041/src/Avalonia.Controls.DataGrid/DataGrid.cs#L628-L652

**Reply by danielmayost** [reply]
I didn't quite understand what you mean, DataGrid adds a property to itself, not to another Control, i don't think that StyledProperty can be added to exists control outside him.

Anyway, in your case this is how it will work:
```
public class ScrollViewerExtensionBehaviour : AvaloniaObject 
    {
        public static readonly AttachedProperty<bool> ShiftWheelScrollsHorizontallyProperty = 
            AvaloniaProperty.RegisterAttached<ScrollViewerExtensionBehaviour, ScrollViewer, bool>("ShiftWheelScrollsHorizontally");

        public static bool GetShiftWheelScrollsHorizontally(ScrollViewer scrollViewer)
        {
            return scrollViewer.GetValue(ShiftWheelScrollsHorizontallyProperty);
        }

        public static void SetShiftWheelScrollsHorizontally(ScrollViewer scrollViewer, bool value)
        {
            scrollViewer.SetValue(ShiftWheelScrollsHorizontallyProperty, value);
        }
    }
```

**Reply by maxkatz6** [reply]
You defined property on the ScrollViewerExtensionBehaviour, but attempt to use it on ScrollViewer.
As mentioned above, you need attached property.

**Reply by KuroiRoy** [reply]
Then it should be like this right?
``` csharp
public static readonly AttachedProperty<bool> ShiftWheelScrollsHorizontallyProperty =
    AvaloniaProperty.RegisterAttached<ScrollViewer, bool>(nameof(ShiftWheelScrollsHorizontally), typeof(ScrollViewerExtensionBehaviour));
```
``` xaml
<ScrollViewer avaloniaUi:ScrollViewerExtensionBehaviour.ShiftWheelScrollsHorizontally="True"></ScrollViewer>
```
But that just gives me another error:
`Import.axaml(31, 10): [XAMLIL] Unable to resolve suitable regular or attached property ShiftWheelScrollsHorizontally on type FinanceTracker:FinanceTracker.Extensions.AvaloniaUI.ScrollViewerExtensionBehaviour Line 31, position 10.`

**Reply by maxkatz6** [reply]
Do you have GetShiftWheelScrollsHorizontally and SetShiftWheelScrollsHorizontally methods?
See https://docs.avaloniaui.net/docs/data-binding/creating-and-binding-attached-properties

**Reply by KuroiRoy** [reply]
Sorry, yeah, that was it. Thank you both very much for helping me understand!
